### PR TITLE
Bump etcd to 3.0.14 and switch to v3 API in etcd.

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #20504

**Release note**:

```release-note
Switch default etcd version to 3.0.14.
Switch default storage backend flag in apiserver to `etcd3` mode.
```

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36229)

### CODE REVIEW

**Comment by dims** [comment]
@wojtek-t Do we need this for 1.5? (I see a 1.5 milestone)

**Comment by wojtek-t** [comment]
@dims - we asked for exception for it. So for now, let's leave 1.5 for this.

**Comment by timothysc** [comment]
I beg to differ on the release not though..

**Comment by timothysc** [comment]
@wojtek-t now that 1.5 is cut, can we move ahead here?

**Comment by wojtek-t** [comment]
@timothysc - yeah sure; I was OOO yesterday, will try to rebase this today.

**Comment by wojtek-t** [comment]
@timothysc @kubernetes/sig-node-misc - both node presubmits doesn't seem to be flake, but I don't know how to debug them. Any hints?

**Comment by xiang90** [comment]
etcd related changes looks good.

**Comment by wojtek-t** [comment]
OK - I figured out the problems with e2e-node. these were basically due to custom setup of etcd there.
I temporarily changed these to run in etcd v2 mode (one line change) but we should fix that in a following PR.

@xiang90 - if you (or Hongchao) could take care of that it would be great.

**Comment by wojtek-t** [comment]
@mml - I pushed the "alpha" image for now to GCE. I will built the regular "3.0.14" image once you push all the changes/fixes from you local branch (this --force-cluster-new etc.).

**Comment by hongchaodeng** [comment]
>  these were basically due to custom setup of etcd there.

@wojtek-t  Do you know what's special about the setup there?

**Comment by wojtek-t** [comment]
@hongchaodeng - this is a custom in-memory setup - see this file:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/services/etcd.go

**Comment by wojtek-t** [comment]
@mml - friendly ping

**Comment by timothysc** [comment]
> OK - I figured out the problems with e2e-node. these were basically due to custom setup of etcd there. 

iirc I changed the port settings on the initialization is what was weird on the node-e2es..

File: cluster/saltbase/salt/etcd/etcd.manifest [hunk 1]
```diff
@@ -39,7 +39,7 @@
 "containers":[
     {
     "name": "etcd-container",
-    "image": "gcr.io/google_containers/etcd:{{ pillar.get('etcd_docker_tag', '2.2.1') }}",
+    "image": "gcr.io/google_containers/etcd:{{ pillar.get('etcd_docker_tag', '3.0.14-alpha.1') }}",
```
Comment (timothysc) [inline cluster/saltbase/salt/etcd/etcd.manifest]: -alpha?

**Comment by timothysc** [comment]
only minor comment.  

/lgtm

File: cluster/saltbase/salt/etcd/etcd.manifest (hunk 1 above)
Comment (wojtek-t) [inline cluster/saltbase/salt/etcd/etcd.manifest]: I wanted to be consistent with versioning scheme for k8s itself, where we are doing "-alpha.1 -alpha.2, ... -beta.1, -beta.2, ...".

I don't think it is that important though, so leaving as is (we can change it later if you feel strongly about it).

**Comment by wojtek-t** [comment]
@k82cn kops test this, issue: #IGNORE (Waiting for status to be reported)

**Comment by wojtek-t** [comment]
@k8s-bot kops aws test this, issue: #IGNORE (Waiting for status to be reported)

**Comment by wojtek-t** [comment]
It seems that kops is using some completely different scripts and it is still trying to use 2.2.1 version of etcd.
@zmerlynn @justinsb - can you please suggest how to proceed with this PR then?

**Comment by wojtek-t** [comment]
@zmerlynn - I'm fine with doing it in a separate PR, but we definitely want to do it in this release. The plan is that etcd v3 will be a default in 1.6 (it was supposed to be default in 1.5 even).

**Comment by zmerlynn** [comment]
Switching the default for the standard deployments via hard flag change and switching the defaults for *all* deployments are two different things, and there was no obvious change announced by code for this flag. Not every deployment is explicitly setting --storage-backend, and you're imposing a "flag tax" to even get back to a functional apiserver for those deployments. It should be warned first.

**Comment by justinsb** [comment]
I believe the plan was to switch GCE/GKE kube-up, so that GCE/GKE could figure out all the real-world upgrade snafus going from etcd2 -> etcd3.  Once the upgrade procedure has succeeded a few thousand times in the wild on GCE/GKE, then we should think about encouraging others to follow.

**Comment by timothysc** [comment]
When did kops become a blocking test to the submit-queue @justinsb ?

**Comment by zmerlynn** [comment]
@timothysc: It was re-added yesterday (I've been attempting to make it blocking for a while now, and it didn't stick because of a few issues, but the PR builder is holding on now).

**Comment by wojtek-t** [comment]
> I believe the plan was to switch GCE/GKE kube-up, so that GCE/GKE could figure out all the real-world upgrade snafus going from etcd2 -> etcd3. Once the upgrade procedure has succeeded a few thousand times in the wild on GCE/GKE, then we should think about encouraging others to follow.

@justinsb - this has already happened. We have done the successful migration on more thousands of clusters. We are continuing those tests, but it seems generally ready to bump

**Comment by justinsb** [comment]
@wojtek-t real-world or simulated?

**Comment by timothysc** [comment]
>  It was re-added yesterday (I've been attempting to make it blocking for a while now, and it didn't stick because of a few issues, but the PR builder is holding on now) 

why?  A small number of people support that deployment method... The idea that the whole community pay the kops-tax is a hard pill for me to swallow.

**Comment by wojtek-t** [comment]
> @wojtek-t real-world or simulated?

real-world

**Comment by justinsb** [comment]
@timothysc it turns out the kops builder was catching problems where we were making assumptions that were only valid on GCE & kube-up.  So actually you benefit from this if you're not using kube-up.

Now of course we have a few teething problems, so bear with us on those.

But in this particular case, this is a real test failure.  This is highlighting that real-world installations using etcd2 will break.  And we're having a discussion about whether or not that is intended - so I think this is a good thing, right?

**Comment by justinsb** [comment]
@wojtek-t cool :-)  Are there any docs on the latest procedure?  Do we just need to run this image: https://github.com/kubernetes/kubernetes/tree/master/cluster/images/etcd ?

**Comment by zmerlynn** [comment]
@timothysc: If you want to have the meta-conversation, I suggest a separate bug, or we can do it on the mailing list, just so we're not increasing noise here. But I agree with https://github.com/kubernetes/kubernetes/pull/36229#issuecomment-270395945. (FWIW, GKE is even worse here. It's a deployment method that isn't even open.)

**Comment by timothysc** [comment]
@wojtek-t could you please add a release note here.

**Comment by justinsb** [comment]
Spoke to @timothysc and I'm going to tweak kops right now to specify the storage=etcd2 flag, to unblock this PR.

**Comment by wojtek-t** [comment]
@justinsb - thanks a lot for it!

@timothysc - sure, adding

**Comment by zmerlynn** [comment]
The PR is in, but it'll need to filter through `kops-aws-updown` to get
escalated to `kops-aws`, and that'll need to be green before the PR builder
picks it up. I'll kick it later today if no one else has.

**Comment by wojtek-t** [comment]
@zmerlynn - thanks a lot for taking care of it.

**Comment by zmerlynn** [comment]
@k8s-bot aws test this

**Comment by ravilr** [comment]
Please make sure to also include a warning note as suggested by @zmerlynn in the release note. something like:
"Needs etcd3 data migration. if you are using custom cluster install/upgrade procedures (not using kube-up), follow instructions to migrate etcd data storage: <link to the general procedure for etcd3 data migration>"

**Comment by justinsb** [comment]
Could we get some docs on the upgrade procedure also?

**Comment by wojtek-t** [comment]
> Could we get some docs on the upgrade procedure also?

Yes sure - this will come. @hongchaodeng @xiang90 @timothysc @mml  FYI

**Pull request merged**